File: uoo6uoow/Week10/균형잡힌 세상.java
import java.io.*;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

        while (true) {
            String input = reader.readLine();
            if (input.equals(".")) {
                break;
            }

            if (balnace(input)) {
                writer.write("yes\n");
            } else {
                writer.write("no\n");
            }
        }
        writer.flush();
    }

    public static boolean balnace(String input) {
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < input.length(); i++) {
            char ch = input.charAt(i);

            if (ch == '(' || ch == '[') {
                stack.push(ch);
            } else if (ch == ')') {
                if (stack.isEmpty() || stack.peek() != '(') {
                    return false;  // 스택이 비었거나 짝이 안 맞을 경우
                }
                stack.pop();
            } else if (ch == ']') {
                if (stack.isEmpty() || stack.peek() != '[') {
                    return false;  // 스택이 비었거나 짝이 안 맞을 경우
                }
                stack.pop();
            }
        }
        return stack.isEmpty();  // 스택이 비었으면 균형이 맞음
    }
}

File: noeyeyh/Week10/균형잡힌 세상.py
results = []

while True:
  s = input()
    
  if s == '.':
    break

  stack = []
  valid = True 

  for i in s:
    if i == '(' or i == '[':
      stack.append(i)  # 열린 괄호를 스택에 추가
    elif i == ')':
      if stack and stack[-1] == '(':  # 짝이 맞는 경우
        stack.pop()
      else:  # 짝이 맞지 않는 경우
        results.append('no')
        valid = False
        break
    elif i == ']':
      if stack and stack[-1] == '[':  # 짝이 맞는 경우
        stack.pop()
      else:  # 짝이 맞지 않는 경우
        results.append('no')
        valid = False
        break

  if valid:
    if not stack:  # 스택이 비어 있으면 모든 괄호가 닫힌 것
      results.append('yes')
    else:
      results.append('no')


for i in results:
  print(i)
